pontoon/base/static/js/fluent_interface.js

```diff
@@ -1,15 +1,39 @@
 /* global FluentSyntax */
 var fluentParser = new FluentSyntax.FluentParser({ withSpans: false });
+var fluentParserOriginal = new FluentSyntax.FluentParser({ withSpans: false });
 var fluentSerializer = new FluentSyntax.FluentSerializer();
 
+/*
+ * VariantLists behave the same as "selector-less SelectExpressions" in Pontoon.
+ * Overriding the parseEntry() method appliess that logic to the AST, too. That
+ * allows us to use the same code for both, VariantLists and SelectExpressions.
+ */
+fluentParser.parseEntry = function(source, useDefaultParser) {
```

Comment:
The fact that you need to override the official implementation is a red flag. Either the design of `VariantList` is wrong, or there has to be a better way to do this. This approach creates a lot of technical debt. I'll be more than happy to help you find a better solution :)

Looking at the source code, in [line 642](https://github.com/mathjazz/pontoon/blob/d206ae3940fe22b4cbdb3f493da05284e8c82c61/pontoon/base/static/js/fluent_interface.js#L642-L645) there's this snippet:

```js
// Value
else if (ast.value) {
    value = renderOriginalElements(ast.value.elements, 'Value');
}
```

You could check if `ast.value` is of type `Pattern` or `VariantList` here and pick between `renderOriginalElements` and, say, `renderOriginalVariants`. Or abstract this check away into a new `renderOriginalValue` method. There might be some duplication regarding the code which renders variants of `SelectExpressions` and variants of `VariantLists`, but not too much, and I'll take it over a custom `parseEntry` any day :)

Reply:
Perhaps I should have stressed that I'm looking at this implementation as a way to postpone yet another refactor of `fluent_interface.js`, and try to do it after or even as part of the rewrite in React.

Reply:
I might be missing something. I'm not suggesting a full refactor. Is my suggestion about line 642 wrong? I really think implementing `VariantList` the proper way at this time shouldn't be very hard and should make the code simpler.

Reply:
Your suggestion is correct - actually it's what I did on backend (https://github.com/mozilla/pontoon/pull/1026/commits/78b9c355c66f2172f117e338fb9ef97a146882a6), where I replaced the `*_elements` method with the `*_value` method.

I completely agree that this is the right approach in general, but on frontend we have much more code that requires such transformation. Here's why I call it a refactor:

1. There are 4 methods used for different string presentation styles:
* `Pontoon.fluent.renderOriginal()`
* `Pontoon.fluent.renderEditor()`
* `Pontoon.fluent.serializeTranslation()`
* `Pontoon.fluent.getSimplePreview()`

2. There are 8 helper functions that take `element` or `elements` as an argument:
* `isSimpleElement`
* `isPluralElement`
* `renderEditorElement`
* `renderEditorElements`
* `renderOriginalElement`
* `renderOriginalElements`
* `serializeFTLEditorElements`
* `stringifyElements`

3. There are 18 calls to `value.elements`, most of which we can no longer make safely.

4. There are things that I didn't think about yet. :)

I'm not saying that every single one of these will need to be refactored, but if https://github.com/mozilla/pontoon/pull/1026/commits/78b9c355c66f2172f117e338fb9ef97a146882a6 resulted in a 55-line changeset for just 1 of these functions, I'd be surprised if we don't get to a 10-times bigger changeset for the entire `fluent_interface.js`. Changing 500 lines in a file with 1000 lines is hardly not a refactor. :)

I also understand parts of code will bi simpler, but in general we're adding complexity. I don't see any simplification beyond removing the two [ifs](https://github.com/mozilla/pontoon/pull/1026/commits/6904925ccf6b4a4636fd98daa649860936ee174e). On the other hand, the proper way of doing things on the [backend](https://github.com/mozilla/pontoon/commit/78b9c355c66f2172f117e338fb9ef97a146882a6#diff-af5d1b44e5d90a99d8307abe864f4bcb) resulted in 30% more code.

Reply:
I understand that there are many parts here which are affected. It might end up being more work than I say. I still firmly believe that it would be better to do this work now than to start with this workaround and effectively do the work twice. I'd prefer to fix the underlying cause rather than the symptom. Fixes to symptoms tend to accumulate until the code becomes unmaintainable and calls for a refactor :)

I don't know how much `fluent_interface.js` is going to be affected by the translate view re-write. You have better visibility into this and I trust your judgement. I understand that this approach allows us to ship Syntax 0.6 faster. I'm worried that by using this work-around we'll make it harder to add code to `fluent_interface` in the longer run. How soon do you expect the translate view re-write to happen? What is the life expectancy of this work-around?

Reply:
> I'd prefer to fix the underlying cause rather than the symptom. Fixes to symptoms tend to accumulate until the code becomes unmaintainable and calls for a refactor :)

To go back to the true cause, I wonder if we can solve the problem `VariantLists` are solving by making a less radical change in the AST (i.e. that will shorten the upper list of things that need to be refactored). I don't know enough about `VariantLists` and I fully trust your judgement, so a simple Yes/No answer would be sufficient. :)

> How soon do you expect the translate view re-write to happen? What is the life expectancy of this work-around?

The rewrite is ongoing and is expected to be complete in Q3. So I don't think it makes a big difference maintenance-wise if we do the refactor now or then.

What worries me is how much time we spend refactoring `fluent_interface.js`. And given that this particular refactor is caused by a feature that is known to be pretty obscure, I worry that it might not even be the last refactor needed before 1.0.

Can I get some guarantee that no further refactors will be needed before we ship 1.0? :) I will understand if the answer is no, but in this case I believe it's better to wait with the refactor until the very last.

Reply:
> To go back to the true cause, I wonder if we can solve the problem VariantLists are solving by making a less radical change in the AST (i.e. that will shorten the upper list of things that need to be refactored). 

No, I don't think that's possible. They are a different kind of values than Patterns.

> What worries me is how much time we spend refactoring fluent_interface.js. And given that this particular refactor is caused by a feature that is known to be pretty obscure, I worry that it might not even be the last refactor needed before 1.0.

I see. I don't want to make product decisions for Pontoon. My advice is based on what's best for Fluent long-term. I understand that Pontoon might want to choose a more winded path to get to the full support :)

> Can I get some guarantee that no further refactors will be needed before we ship 1.0? :)

No, sorry, I can't guarantee this. There aren't any ground shaking changes planned before 1.0, but this commitment concerns the syntax. AST might be subject to change, still, and depending on the scope of changes, they might require more work in `fluent_interface`.

Which is also why I'm afraid of taking the work-around route right now. If more changes are needed down the pike, the next refactor is going to be bigger and harder. Or, we'll choose to add more work-arounds, which is how technical debt is created.

That said, I see your point about this being an obscure feature of Fluent, rarely used in Pontoon right. For this reason, it makes sense to go with the work-around. I think what I'm really afraid is that we won't be able to fully (and properly) support Fluent 1.0 before the translate view re-write is complete. Are these two related? Is Fluent 1.0 an explicit goal for the rewrite? Or is it independent (but impacted by it, as we see).

Reply:
Sounds good, @mathjazz, thanks.

Reply:
OK. I believe the PR is ready for a review then.